src/main/java/com/flagsmith/flagengine/segments/SegmentEvaluator.java

```diff
@@ -1,196 +1,212 @@
+          break;
+        default:
+          return false;
+      }
     }
 
-    return Boolean.FALSE;
+    return isMatch && rule.getRules().stream()
+        .allMatch((subRule) -> contextMatchesRule(context, subRule, segmentKey));
   }
 
-  /**
-   * Evaluate whether the trait match the rule from segment.
-   *
-   * @param identityTraits Traits to match against.
-   * @param rule Rule from segments to evaluate with.
-   * @param segmentId Segment ID (for hashing)
-   * @param identityId Identity ID (for hashing)
-   */
-  private static Boolean traitsMatchSegmentRule(List<? extends TraitModel> identityTraits,
-                                                SegmentRuleModel rule,
-                                                Integer segmentId, String identityId) {
-    Boolean matchingCondition = Boolean.TRUE;
-
-    if (rule.getConditions() != null && rule.getConditions().size() > 0) {
-      List<Boolean> conditionEvaluations = rule.getConditions().stream()
-          .map((condition) -> traitsMatchSegmentCondition(identityTraits, condition, segmentId,
-              identityId))
-          .collect(Collectors.toList());
-
-      matchingCondition = rule.matchingFunction(
-          conditionEvaluations.stream()
-      );
-    }
+  private static Boolean contextMatchesCondition(
+      EvaluationContext context,
+      SegmentCondition condition,
+      String segmentKey) {
+    Object contextValue = getContextValue(context, condition.getProperty());
+    Object conditionValue = condition.getValue();
+    SegmentConditions operator = condition.getOperator();
 
-    List<SegmentRuleModel> rules = rule.getRules();
+    switch (operator) {
+      case IN:
+        List<String> conditionList = new ArrayList<>();
+
+        if (conditionValue instanceof List) {
+          List<?> maybeConditionList = (List<?>) conditionValue;
+          conditionList = maybeConditionList.stream()
+              .filter(String.class::isInstance)
+              .map(Object::toString)
+              .collect(Collectors.toList());
+        } else if (conditionValue instanceof String) {
+          String stringConditionValue = (String) conditionValue;
+          try {
+            // Try parsing a JSON list first
+            conditionList = mapper.readValue(
+                stringConditionValue, stringListTypeRef);
+          } catch (IOException e) {
+            // As a fallback, split by comma
+            conditionList = Arrays.asList(stringConditionValue.split(","));
+          }
+        }
 
-    if (rules != null) {
-      matchingCondition = matchingCondition && rules.stream()
-          .allMatch((segmentRule) -> traitsMatchSegmentRule(
-              identityTraits,
-              segmentRule,
-              segmentId,
-              identityId
-          ));
-    }
+        if (!(contextValue instanceof Boolean) && contextValue != null) {
+          contextValue = String.valueOf(contextValue);
+        }
```

Comment:
I don't understand why booleans are left out of the party here. 😞  Will `.contains` below compare `"true" == true` / `"false" == false`? Sorry I don't follow. Our [whole boolean thing](https://github.com/Flagsmith/flagsmith-engine/blob/789d737de3c6ec9594632ccbada5c96e8622785b/flag_engine/utils/types.py#L31) is already a bit of a party pooper.